subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/plugins/JavaGradlePluginPlugin.java

```diff
@@ -182,103 +188,86 @@ private void configureTestKit(Project project, GradlePluginDevelopmentExtension
+            CopySpec copyPluginDescriptors = task.getRootSpec().addChild();
+            copyPluginDescriptors.into("META-INF/gradle-plugins");
+            copyPluginDescriptors.from(generatePluginDescriptors);
         });
     }
 
-    private void validatePluginDeclarations(Project project, final GradlePluginDevelopmentExtension extension) {
-        project.afterEvaluate(new Action<Project>() {
-            @Override
-            public void execute(Project project) {
-                for (PluginDeclaration declaration : extension.getPlugins()) {
-                    if (declaration.getId() == null) {
-                        throw new IllegalArgumentException(String.format(DECLARATION_MISSING_ID_MESSAGE, declaration.getName()));
-                    }
-                    if (declaration.getImplementationClass() == null) {
-                        throw new IllegalArgumentException(String.format(DECLARATION_MISSING_IMPLEMENTATION_MESSAGE, declaration.getName()));
-                    }
+    private void validatePluginDeclarations(Project project, GradlePluginDevelopmentExtension extension) {
+        project.afterEvaluate(evaluatedProject -> {
+            for (PluginDeclaration declaration : extension.getPlugins()) {
+                if (declaration.getId() == null) {
+                    throw new IllegalArgumentException(String.format(DECLARATION_MISSING_ID_MESSAGE, declaration.getName()));
+                }
+                if (declaration.getImplementationClass() == null) {
+                    throw new IllegalArgumentException(String.format(DECLARATION_MISSING_IMPLEMENTATION_MESSAGE, declaration.getName()));
                 }
             }
         });
     }
 
-    private void configureTaskPropertiesValidation(final Project project) {
-        final SourceSet mainSourceSet = project.getConvention().getPlugin(JavaPluginConvention.class).getSourceSets().getByName(SourceSet.MAIN_SOURCE_SET_NAME);
-        final TaskProvider<ValidateTaskProperties> validator = project.getTasks().register(VALIDATE_TASK_PROPERTIES_TASK_NAME, ValidateTaskProperties.class, new Action<ValidateTaskProperties>() {
-            @Override
-            public void execute(ValidateTaskProperties validator) {
-                validator.setGroup(PLUGIN_DEVELOPMENT_GROUP);
-                validator.setDescription(VALIDATE_TASK_PROPERTIES_TASK_DESCRIPTION);
+    private void configurePluginValidation(Project project) {
+        SourceSet mainSourceSet = project.getConvention().getPlugin(JavaPluginConvention.class).getSourceSets().getByName(SourceSet.MAIN_SOURCE_SET_NAME);
+        TaskProvider<Task> naggerTask = project.getTasks().register("nagAboutValidateTaskProperties", task -> {
+            task.setGroup(PLUGIN_DEVELOPMENT_GROUP);
+            task.setDescription("Nag about task " + VALIDATE_TASK_PROPERTIES_TASK_NAME + " replaced by " + VALIDATE_PLUGIN_TASK_NAME);
+            task.doFirst(taskWithAction -> nagAboutDeprecatedValidateTaskPropertiesTask());
```

Comment:
Why don't you add the nagging as a task action of `ValidateTaskProperties`?

Moreover, I suppose we shouldn't use Lambdas as task actions, given that we can't track their implementation. It doesn't really matter here since the task doesn't declare any inputs or outputs, though we probably should be consistent.

Reply:
FTR, `ValidateTaskProperties` won't be executed if `ValidatePlugins` fails, and the two tasks can't both `dependsOn` and `finalizedBy` each other at the same time.

As to the lambda: I replaced it. We should consider deprecating their use, so I added https://github.com/gradle/gradle/issues/10751.